File: demo/xiaoquan/skills/scheduler_mgr/scripts/_tasks_store.py
import argparse
import json
import sys
import time
from pathlib import Path
from typing import Any, Dict, List


TASKS_PATH = Path("/workspace/cron/tasks.json")


def _load_store() -> Dict[str, Any]:
    if not TASKS_PATH.exists():
        return {"version": 1, "jobs": []}
    try:
        data = json.loads(TASKS_PATH.read_text(encoding="utf-8"))
        if not isinstance(data, dict):
            raise ValueError("tasks.json root is not an object")
        jobs = data.get("jobs")
        if not isinstance(jobs, list):
            raise ValueError("tasks.json.jobs is not an array")
        return {"version": int(data.get("version", 1)), "jobs": list(jobs)}
    except Exception as exc:  # noqa: BLE001
        raise RuntimeError(f"failed to load tasks.json: {exc}") from exc


def _dump_store(store: Dict[str, Any]) -> None:
    TASKS_PATH.parent.mkdir(parents=True, exist_ok=True)
    content = json.dumps(store, ensure_ascii=False, indent=2)
    tmp = TASKS_PATH.with_suffix(".tmp")
    tmp.write_text(content, encoding="utf-8")
    tmp.rename(TASKS_PATH)


def _now_ms() -> int:
    return int(time.time() * 1000)


def list_jobs() -> Dict[str, Any]:
    store = _load_store()
    jobs: List[Dict[str, Any]] = store.get("jobs", [])
    return {
        "errcode": 0,
        "errmsg": "success",
        "data": {"total": len(jobs), "jobs": jobs},
    }


def create_job(
    name: str,
    schedule_kind: str,
    routing_key: str,
    message: str,
    expr: str | None = None,
    tz: str | None = None,
    at_ms: int | None = None,
    every_ms: int | None = None,
    delete_after_run: bool = False,
) -> Dict[str, Any]:
    import uuid

    if schedule_kind not in {"at", "every", "cron"}:
        return {
            "errcode": 1,
            "errmsg": f"无效的 schedule.kind：{schedule_kind}\n建议：使用 at/every/cron 之一。",
            "data": {},
        }

    if schedule_kind == "cron" and not expr:
        return {
            "errcode": 1,
            "errmsg": "cron 任务必须提供 expr 字段。\n建议：例如 '0 9 * * 1-5'。",
            "data": {},
        }

    if schedule_kind == "at" and at_ms is None:
        return {
            "errcode": 1,
            "errmsg": "at 任务必须提供 at_ms（Unix 毫秒时间戳）。",
            "data": {},
        }

    if schedule_kind == "every" and every_ms is None:
        return {
            "errcode": 1,
            "errmsg": "every 任务必须提供 every_ms（间隔毫秒数）。",
            "data": {},
        }

    store = _load_store()
    now_ms = _now_ms()

    job_id = f"job-{uuid.uuid4().hex[:8]}"
    schedule: Dict[str, Any] = {
        "kind": schedule_kind,
        "at_ms": at_ms if schedule_kind == "at" else None,
        "every_ms": every_ms if schedule_kind == "every" else None,
        "expr": expr if schedule_kind == "cron" else None,
        "tz": tz if schedule_kind == "cron" else None,
    }

    job: Dict[str, Any] = {
        "id": job_id,
        "name": name,
        "enabled": True,
        "schedule": schedule,
        "payload": {
            "routing_key": routing_key,
            "message": message,
        },
        "state": {
            "next_run_at_ms": None,
            "last_run_at_ms": None,
            "last_status": None,
            "last_error": None,
        },
        "created_at_ms": now_ms,
        "updated_at_ms": now_ms,
        "delete_after_run": delete_after_run,
    }

    store["jobs"].append(job)
    _dump_store(store)

    return {
        "errcode": 0,
        "errmsg": "success",
        "data": {
            "action": "created",
            "job_id": job_id,
            "name": name,
            "schedule_kind": schedule_kind,
        },
    }


def delete_job(job_id: str) -> Dict[str, Any]:
    store = _load_store()
    jobs: List[Dict[str, Any]] = store.get("jobs", [])
    before = len(jobs)
    jobs = [j for j in jobs if j.get("id") != job_id]

    if len(jobs) == before:
        return {
            "errcode": 1,
            "errmsg": f"未找到要删除的任务：{job_id}",
            "data": {},
        }

    store["jobs"] = jobs
    _dump_store(store)
    return {
        "errcode": 0,
        "errmsg": "success",
        "data": {"action": "deleted", "job_id": job_id},
    }


def update_job(
    job_id: str,
    name: str | None = None,
    enabled: bool | None = None,
    routing_key: str | None = None,
    message: str | None = None,
    schedule_kind: str | None = None,
    expr: str | None = None,
    tz: str | None = None,
    at_ms: int | None = None,
    every_ms: int | None = None,
    delete_after_run: bool | None = None,
) -> Dict[str, Any]:
    store = _load_store()
    jobs: List[Dict[str, Any]] = store.get("jobs", [])
    target = None
    for j in jobs:
        if j.get("id") == job_id:
            target = j
            break
    if target is None:
        return {
            "errcode": 1,
            "errmsg": f"未找到要更新的任务：{job_id}",
            "data": {},
        }

    changed = False
    if name is not None:
        target["name"] = name
        changed = True
    if enabled is not None:
        target["enabled"] = enabled
        changed = True
    if delete_after_run is not None:
        target["delete_after_run"] = delete_after_run
        changed = True

    payload = target.setdefault("payload", {})
    if routing_key is not None:
        payload["routing_key"] = routing_key
        changed = True
    if message is not None:
        payload["message"] = message
        changed = True

    schedule = target.setdefault("schedule", {})
    if schedule_kind is not None:
        schedule["kind"] = schedule_kind
        changed = True
    if expr is not None:
        schedule["expr"] = expr
        changed = True
    if tz is not None:
        schedule["tz"] = tz
        changed = True
    if at_ms is not None:
        schedule["at_ms"] = at_ms
        changed = True
    if every_ms is not None:
        schedule["every_ms"] = every_ms
        changed = True

    if not changed:
        return {
            "errcode": 0,
            "errmsg": "success",
            "data": {"action": "noop", "job_id": job_id},
        }

    target_state = target.setdefault("state", {})
    # 修改 schedule 后，让 CronService 重新计算 next_run_at_ms
    target_state["next_run_at_ms"] = None
    target_state.setdefault("last_run_at_ms", None)
    target_state.setdefault("last_status", None)
    target_state.setdefault("last_error", None)

    target["updated_at_ms"] = _now_ms()
    store["jobs"] = jobs
    _dump_store(store)

    return {
        "errcode": 0,
        "errmsg": "success",
        "data": {"action": "updated", "job_id": job_id},
    }


def _main() -> None:
    parser = argparse.ArgumentParser(description="Low-level tasks.json helper.")
    parser.add_argument("action", choices=["list", "create", "delete", "update"])
    parser.add_argument("--name")
    parser.add_argument("--job_id")
    parser.add_argument("--routing_key")
    parser.add_argument("--message")
    parser.add_argument("--schedule_kind")
    parser.add_argument("--expr")
    parser.add_argument("--tz")
    parser.add_argument("--at_ms", type=int)
    parser.add_argument("--every_ms", type=int)
    parser.add_argument("--enabled", type=str)
    parser.add_argument("--delete_after_run", type=str)
    args = parser.parse_args()

    try:
        if args.action == "list":
            result = list_jobs()
        elif args.action == "create":
            result = create_job(
                name=args.name or "未命名任务",
                schedule_kind=args.schedule_kind or "cron",
                routing_key=args.routing_key or "",
                message=args.message or "",
                expr=args.expr,
                tz=args.tz,
                at_ms=args.at_ms,
                every_ms=args.every_ms,
                delete_after_run=(
                    str(args.delete_after_run).lower() == "true"
                    if args.delete_after_run is not None
                    else False
                ),
            )
        elif args.action == "delete":
            if not args.job_id:
                raise ValueError("delete 操作必须提供 --job_id")
            result = delete_job(args.job_id)
        else:  # update
            if not args.job_id:
                raise ValueError("update 操作必须提供 --job_id")
            enabled = None
            if args.enabled is not None:
                enabled = args.enabled.lower() == "true"
            delete_after_run = None
            if args.delete_after_run is not None:
                delete_after_run = args.delete_after_run.lower() == "true"
            result = update_job(
                job_id=args.job_id,
                name=args.name,
                enabled=enabled,
                routing_key=args.routing_key,
                message=args.message,
                schedule_kind=args.schedule_kind,
                expr=args.expr,
                tz=args.tz,
                at_ms=args.at_ms,
                every_ms=args.every_ms,
                delete_after_run=delete_after_run,
            )
    except Exception as exc:  # noqa: BLE001
        result = {
            "errcode": 1,
            "errmsg": f"内部错误：{exc}",
            "data": {},
        }

    sys.stdout.write(json.dumps(result, ensure_ascii=False))
    sys.stdout.flush()


if __name__ == "__main__":
    _main()

File: demo/xiaoquan/src/cron/tasks-store.js
/**
 * tasks-store.js — cron/tasks.json 原子读写 API
 *
 * 对齐 Python xiaopaw-team 的 tasks_store.py：
 * - write-then-rename 原子写入，CronService 通过 mtime 检测热重载
 * - scheduleWake: 去重（同 routingKey+message 的未到期 at job 复用）
 * - 锁：用 fs.open(lockPath, 'wx') 实现进程间文件锁（retry）
 */

import fs from 'fs'
import path from 'path'
import {randomBytes} from 'crypto'
import {setTimeout as sleep} from 'timers/promises'

function nowMs() {
  return Date.now()
}

function genJobId() {
  return 'job-' + randomBytes(4).toString('hex')
}

function lockPath(tasksPath) {
  return tasksPath + '.lock'
}

async function acquireLock(tasksPath, {timeoutMs = 5000, retryMs = 20} = {}) {
  fs.mkdirSync(path.dirname(tasksPath), {recursive: true})
  const lp = lockPath(tasksPath)
  const deadline = Date.now() + timeoutMs
  while (Date.now() < deadline) {
    try {
      const fd = fs.openSync(lp, 'wx')
      fs.closeSync(fd)
      return true
    } catch {
      await sleep(retryMs)
    }
  }
  throw new Error(`Could not acquire lock on ${lp} after ${timeoutMs}ms`)
}

function releaseLock(tasksPath) {
  try { fs.unlinkSync(lockPath(tasksPath)) } catch {}
}

function loadStore(tasksPath) {
  if (!fs.existsSync(tasksPath)) return {version: 1, jobs: []}
  try {
    const raw = fs.readFileSync(tasksPath, 'utf-8') || '{"version":1,"jobs":[]}'
    const data = JSON.parse(raw)
    if (Array.isArray(data)) return {version: 1, jobs: data}
    return {version: Number(data.version || 1), jobs: Array.isArray(data.jobs) ? data.jobs : []}
  } catch {
    return {version: 1, jobs: []}
  }
}

function dumpStore(tasksPath, store) {
  const dir = path.dirname(tasksPath)
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, {recursive: true})
  const tmp = tasksPath + '.tmp'
  fs.writeFileSync(tmp, JSON.stringify(store, null, 2), 'utf-8')
  fs.renameSync(tmp, tasksPath)
}

export async function createJob(tasksPath, {
  name,
  routingKey,
  message,
  scheduleKind = 'at',
  atMs = null,
  everyMs = null,
  expr = null,
  tz = null,
  deleteAfterRun = true,
} = {}) {
  if (!['at', 'every', 'cron'].includes(scheduleKind)) throw new Error(`invalid scheduleKind: ${scheduleKind}`)
  if (scheduleKind === 'at' && atMs == null) atMs = nowMs() + 1000
  if (scheduleKind === 'every' && everyMs == null) throw new Error('everyMs required for every')
  if (scheduleKind === 'cron' && !expr) throw new Error('expr required for cron')

  await acquireLock(tasksPath)
  try {
    const store = loadStore(tasksPath)
    const now = nowMs()
    const jobId = genJobId()
    store.jobs.push({
      id: jobId,
      name,
      enabled: true,
      schedule: {kind: scheduleKind, atMs: scheduleKind === 'at' ? atMs : null,
        everyMs: scheduleKind === 'every' ? everyMs : null,
        expr: scheduleKind === 'cron' ? expr : null, tz: scheduleKind === 'cron' ? tz : null},
      payload: {routingKey, message},
      state: {nextRunAtMs: null, lastRunAtMs: null, lastStatus: null, lastError: null},
      createdAtMs: now,
      updatedAtMs: now,
      deleteAfterRun: Boolean(deleteAfterRun),
    })
    dumpStore(tasksPath, store)
    return jobId
  } finally {
    releaseLock(tasksPath)
  }
}

export async function scheduleWake(tasksPath, {role, reason = 'new_mail', delayMs = 1000, projectId = ''} = {}) {
  const tag = projectId ? `:${projectId}` : ''
  const targetMsg = `__wake__:${reason}${tag}`
  const targetRk = `team:${role}`

  // 去重：查找同 routingKey+message 且未到期的 at job
  await acquireLock(tasksPath)
  try {
    const now = nowMs()
    const store = loadStore(tasksPath)
    for (const j of store.jobs) {
      if (j.payload?.routingKey === targetRk &&
          j.payload?.message === targetMsg &&
          j.schedule?.kind === 'at' &&
          j.deleteAfterRun &&
          (j.schedule?.atMs || 0) > now) {
        return j.id  // 复用已有 pending wake
      }
    }
    // 无重复，创建新 job
    const jobId = genJobId()
    const now2 = nowMs()
    store.jobs.push({
      id: jobId,
      name: `wake-${role}-${reason}-${now2}`,
      enabled: true,
      schedule: {kind: 'at', atMs: now2 + delayMs, everyMs: null, expr: null, tz: null},
      payload: {routingKey: targetRk, message: targetMsg},
      state: {nextRunAtMs: null, lastRunAtMs: null, lastStatus: null, lastError: null},
      createdAtMs: now2,
      updatedAtMs: now2,
      deleteAfterRun: true,
    })
    dumpStore(tasksPath, store)
    return jobId
  } finally {
    releaseLock(tasksPath)
  }
}

export function listJobs(tasksPath) {
  return loadStore(tasksPath).jobs
}

export async function deleteJob(tasksPath, {jobId}) {
  await acquireLock(tasksPath)
  try {
    const store = loadStore(tasksPath)
    const before = store.jobs.length
    store.jobs = store.jobs.filter(j => j.id !== jobId)
    if (store.jobs.length === before) return false
    dumpStore(tasksPath, store)
    return true
  } finally {
    releaseLock(tasksPath)
  }
}
